theme: switch to vanilla grid, format footer

medubelko opened the pull request:
General:

- Switch to more modern code reuse method for SCSS.
- Make content text full-width.

Layout:

- Use [Vanilla's docs layout](https://vanillaframework.io/docs/layouts/documentation) for main body. Not all of the grid areas are in use because of limitations with how the main content composition, but it's ready for further development. The `header` and `title` elements are ignored. These should be followed up on later.
- Modify Vanilla's grid so it accords with documentation needs, specifically a single-column, centered content block.
- Update the base page template to be more modular; move the containing tags into the template files themselves.
- Move the last-updated date and bug report link to top of the content block. This is placeholder positioning until we are able to intercept the main content composition.

For DOCPR-1685:

- Use general footer layout from Vanilla.
- Include a sparing set of lines.
- Implement correct copyright and license notices, provide additional
  config options for both.

medubelko committed:
theme: switch to vanilla grid, format footer

General:

- Make SCSS stack clearer by delineating between pre- and post-render.
    - Fixes the recent typography settings, which weren't applying.
- Make content text full-width.

Layout:

- Use Vanilla's built-in docs layout for main body. Not all of the
  grid areas are in use because of limitations with how the main content
  composition, but it's ready for further development. The `header` and
  `title` elements are ignored. These should be followed up on later.
- Modify Vanilla's grid so it accords with documentation needs,
  specifically a single-column, centered content block.
- Update the base page template to be more modular; move the containing
  tags into the template files themselves.
- Move the last-updated date and bug report link into a placeholder
  article header. This is placeholder until we are able to intercept
  the main content composition.

For DOCPR-1685:

- Use general footer layout from Vanilla.
- Include a sparing set of lines.
- Implement correct copyright and license notices, provide additional
  config options for both.

izmalk commented on ulwazi/theme/ulwazi/assets/_layout.scss:
```diff
@@ -0,0 +1,22 @@
+//*
+// Set the documentation layout. The main work is to make the article content a column.
+//*
+@use "vanilla-framework" as vanilla;
```
I'll need @nbellol help to validate the SCSS magic here

medubelko replied:
Reverted this; it's not fit for Vanilla's current design.

izmalk commented on docs/conf.py:
```diff
@@ -60,8 +62,8 @@
 
-copyright = "%s CC-BY-SA, %s" % (datetime.date.today().year, author)
-
+copyright = "&copy; %s %s" % (datetime.date.today().year, author)
```
Question: The original version is from our Starterpack: https://github.com/canonical/sphinx-docs-starter-pack/blob/main/docs/conf.py#L63
Can you provide the source for this updated version, please?

P.S. I think it's best to consider the Starter pack to be the source of truth for this. The Ulwazi theme provides a tooling to render it, but the content can and should be adjusted outside the theme.

medubelko replied:
The original isn't a correct copyright notice. :) The format is, pretty much universally:

```
© <year> <holder>
```

Canonical follows this consistently.

> I think it's best to consider the Starter pack to be the source of truth for this

I feel like this is a severe constraint; how do we design our layout for configurable sections if we can't change the configuration?

izmalk replied:
My point is, this repo is for the theme. And theme provides a variable to configure the content that will be shown there. This is not the best place for a discussion of what actually should be there. I'd prefer us using the default starter pack values here for now and discuss whether we can improve them in the starter pack, where, hopefully, more knowledgeable people can review the suggestions and make a conscious decision.
With the current way of deployment for the theme, the conf.py won't be propagated to the final deployment anyway, as it will be overridden by the starter pack. So this conf.py is for testing purposes only. To test how it works, so we can suggest and document necessary configuration parameters later.

medubelko replied:
I'm looking into upstreaming this change.

medubelko replied:
Upstreaming WIP:

https://github.com/canonical/canonical-sphinx/pull/105
https://github.com/canonical/sphinx-docs-starter-pack/pull/444

Implemented in this branch with 03f18ad.

izmalk commented on docs/conf.py:
```diff
@@ -149,6 +149,12 @@
+    # TODO: Add your project's license
+    "license": {
+        "type": "CC-BY-SA",
+        "url": "https://github.com/canonical/ulwazi/blob/main/LICENSE",
```
Todo: Fix a broken link.

izmalk commented on docs/content/test.rst:
```diff
@@ -15,8 +15,10 @@ Test content.
 Nest test
 ---------
 
+test
```
Question: Why do we need this?

medubelko replied:
I put that in there to preview paragraphs. Do you want me to pull it out?

izmalk replied:
We have other pages (e.g., the home page, the cheat sheets) that have paragraphs of text for testing. I'm just wondering why it is needed to add one here.

medubelko replied:
Reverted. 4bab13e

izmalk reviewed (changes requested):
Thank you very much for addressing this.
This is a big PR, changing many things at once. I wish we had separated the changes into smaller, more narrowly scoped PRs. But we can proceed as it is if you want, it will just take some time to review and polish things up.
There are a few opportunities for improvement that I see, namely:
1. The tests are failing at the SCSS properties propagation check.
2. I'm not sure about some of the changes, for example, the Footer contents. What is the source on this Footer? Have you used some of the existing pages or found relevant guidance in Vanilla Framework? IIRC, I've used mostly Canonical.com/data docs sets for the original version as inspiration. See [MAAS](https://canonical.com/maas/docs) and [Spark](https://canonical.com/data/docs/spark/k8s/t-overview) for example.
It would be much easier to review and discuss if you add to the PR description links to the exact reference materials used.
3. The preview shows some problems with rendering now:
<img width="724" height="400" alt="Screenshot from 2025-09-11 11-56-59" src="https://github.com/user-attachments/assets/cc297ed4-4d4c-40ae-8306-7e8a97d113a4" />
<img width="480" height="296" alt="Screenshot from 2025-09-11 12-04-47" src="https://github.com/user-attachments/assets/e1718016-594a-410e-8969-85523b270f06" />
<img width="477" height="273" alt="Screenshot from 2025-09-11 12-04-38" src="https://github.com/user-attachments/assets/adf377e6-ef40-4103-b37f-417c5ed353ae" />

medubelko commented:
> 2. I'm not sure about some of the changes, for example, the Footer contents. What is the source on this Footer? Have you used some of the existing pages or found relevant guidance in Vanilla Framework? IIRC, I've used mostly Canonical.com/data docs sets for the original version as inspiration. See [MAAS](https://canonical.com/maas/docs) and [Spark](https://canonical.com/data/docs/spark/k8s/t-overview) for example.

Great questions! I'll walk you through my logic.

- I'm following recent work we did for the current theme footer. The intent is to keep it as simple and short as possible without adding further complexity that will need dozens of variations per doc set.
    <img width="1340" height="129" alt="image" src="https://github.com/user-attachments/assets/40308405-572b-4a6e-a8c8-73847a4641be" />
- Vanilla has no opinion on what's placed in the footer, and its example is a marketing footer:
    <img width="2215" height="375" alt="image" src="https://github.com/user-attachments/assets/a141c98d-7c4d-4922-88c1-2576e2b67149" />
- Here's snapcraft.io's Vanilla-based footer, live for years. Very busy, quite different from the Vanilla default, and half of it won't apply to docs sets:
    <img width="2286" height="483" alt="image" src="https://github.com/user-attachments/assets/a0e8f609-6763-47cc-bdbe-33977287f7f4" />
    So, we have an opportunity to make a good footer _fit for docs_.
- I suggest that the trademark notice isn't needed unless by fiat from legal. We're not selling a product, and there are many projects that won't be using the Canonical branding. Vanilla's itself doesn't.
- The link to legal is probably needed, but do we want to make that a default for everyone using the theme?
- MAAS and Sparks don't contain correct copyright notices. Copyrights are held by owners, not licenses. The license should be separate statement.

- The bug issue/report link shouldn't be buried in the footer. This has been a clear-cut pattern in our docs for a while.

Maybe I'm misunderstanding our intentions. As a _theme_ that others outside the org might use, do we want to be pushing Canonical-specific decisions to the base template? All of the Canonical-specific legal info will need to be reconsidered for CODA, for example.

I could split the footer into two columns like in Vanilla's example. To me that feels to be the most recognisable and consistent part of our footers.

> It would be much easier to review and discuss if you add to the PR description links to the exact reference materials used.

I'll include those next time, but there are cases where we'll have to come up with new things based on our own judgement and design, no?

> 3. The preview shows some problems with rendering now:

Can you share a wider context for the second screenshot? I can't tell what viewport size and scale is leading to that.

medubelko commented:
Side nav error fixed in 57b3679.

Test error fixed in 724753a.

izmalk commented:
> I'm following recent work we did for the current theme footer. 

Well, would you agree, that both examples that you've provided are looking better than what can be seen in the preview of this branch?
<img width="1916" height="484" alt="Screenshot from 2025-09-12 09-17-10" src="https://github.com/user-attachments/assets/501cb4b4-5afc-4d8f-96fa-09eecafc7528" />
I find your example of the marketing footer rather close to what we've been aiming for so far. Maybe with some other links or even configurable set of links. I don't have a strong opinion of what exact links and text should be in the footer. But it needs to be nicely designed.

To be clear - the usability of the Ulwazi theme outside Canonical docs is a long-term goal, but out of scope for this particular epic and cycle. I don't really care much about the trademark notice being there, to be honest, but I can't see it hurting anyone at this point. I consider it to be a good "lorem ipsum" text for us at this point. Maybe we'll decide to replace it. But having a huge and empty footer doesn't look great. 

> there are cases where we'll have to come up with new things based on our own judgement and design, no?

Yes, of course, and these cases often require some explanation/justification or references to inspirational materials. That helps to speed up the review process.

> Can you share a wider context for the second screenshot? I can't tell what viewport size and scale is leading to that.

I'm not sure what you mean. All our pages, regardless of scale, have these two lines up top. I was merely wondering why would they be up top and occupy the most expensive real estate on the screen. Seems out of place to me. Here is a full screen of a page:
<img width="2557" height="1327" alt="Screenshot from 2025-09-12 09-31-41" src="https://github.com/user-attachments/assets/20eb9eef-16c2-4052-819d-16db645e9e5a" />

medubelko committed:
theme: make the body a grid and the footer sticky

medubelko committed:
theme: revert change to test doc

medubelko committed:
theme: use two-column footer layout

medubelko commented:
Rebased branch with latest from main.

@izmalk 

Sorry for missing the earlier display issue; it should be fixed now.

I've incorporated a two-column footer like in the marketing example. Are there any other items you'd like to add to the nav list?

medubelko committed:
theme: fix rtd flyout causing space under footer

medubelko committed:
theme: align copyright statement

medubelko committed:
theme: add modular copyright and license

nbellol reviewed (approved):
It looks good to me, all works correctly!

Requires a rebase to solve the conflicts!

Maybe for a Future PR we can check the responsiveness since the grid changes the way the pages render on small screens

<img width="1926" height="1658" alt="image" src="https://github.com/user-attachments/assets/5f20c80a-547e-4c51-b201-68e2c1b6dc46" />

izmalk reviewed (changes requested):
It looks better now. I have to ask for another rebase and a couple non-blocking suggestion:
1. Todo: We definitely need to rebase it again to see how it works with the new Nav menu made by Nicolas.
2. Note: I'd prefer to have the contents of the Footer configurable from Conf.py, rather than relying on people using custom HTML templates. For that, I'd suggest making the first line optional, and adding a possibility to add more lines with custom text and link. But we can add it later in a separate Footer.
3. Question: Since we are touching the grid, can we later use it to align better the horizontal Nav Menu and page's content and title? See screenshot.

<img width="508" height="417" alt="Image" src="https://github.com/user-attachments/assets/36338867-b402-447d-8eef-933730a8bba3" />

medubelko committed:
Merge branch 'main' into docpr-1685-switch-to-vanilla-grid-format-footer

medubelko committed:
theme: fix nav element alignment and size in grid

medubelko committed:
theme: harmonise navigation bars with grid

medubelko committed:
theme: make footer customisable

The footer is ordered as:

- Product name
- License
- Custom entry 1
- Custom entry 2...

Add two config settings for the footer:

- `footer.product`, a bool for whether to include the product name.
- `footer.entries`, a list of all custom entries.

medubelko committed:
theme: make license toggleable in the footer

medubelko commented:
@nbellol

Rebased. e37dc82  also fixes a small bug on smaller screens in the base TOC.

Before | After
-|-
<img width="66%" alt="Screen Shot 2025-10-14 at 21 33 35" src="https://github.com/user-attachments/assets/ee5159f0-2dfc-454b-9f35-7a0e954f3d66" /> | <img width="66%" alt="Screen Shot 2025-10-14 at 21 32 58" src="https://github.com/user-attachments/assets/2e3a6497-290b-4088-a5df-270639580e0e" />

@izmalk 

7623dd5 puts the nav bars on the grid.

15bfecf and 9abca6c make the footer customisable, but it's rudimentary. There isn't an easy way to allow for linkable text as config settings, so `footer.entries` simply copies text + HTML that the author provides. The product name and the license are special cases, so they're toggleable.

izmalk commented on docs/conf.py:
```diff
@@ -31,13 +31,15 @@
 #
 # TODO: To include a version number, add it here (hardcoded or automated).
-#
+version = "beta"
```
Question: Why do we set this, and why is it "beta"?
It's not even being used, since RTD uses environment variable instead AFAIK.

medubelko replied:
It's one of the few [default global vars](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-version) in Sphinx. I'm using this value as a placeholder, because the intent is for the footer to contain a line like `Charmed Apache Spark v0.0.4`.

For example, in Starcraft, we assign the version from the project source, although we use a different var name for legacy reasons. https://github.com/canonical/snapcraft/blob/a634d1f08ecd1241023c78462b10fbdf9761007e/docs/conf.py#L29. Our users could do something similar.

I think it's good to use the var this way, and we should do this for the eventual title element, too. Otherwise, non-RTD users won't have a way of displaying which version the docs correspond to. What do you think?

izmalk replied:
I like the idea of adding it to the title to differentiate the version.
I'd even double-down and make it check the URL scheme, and only add the version to the title if the URL scheme supports multi-versioning. Or at least have it so you need to explicitly enable the version in config (with the default to be off).
The main problem I see, is that it's out of scope for this PR. I saw the footer, but I'm not sure it's needed in footer, and in any case, on RTD this variable is overridden by an environment variable, so we are not even using it other than for local builds.
As I said, it's one of my concerns that can be polished out next cycle when we actually get to refactoring the code.

medubelko replied:
Sounds good, I'd be happy to revisit later.

izmalk commented on docs/conf.py:
```diff
@@ -31,13 +31,15 @@
 
 
-# Copyright string; shown at the bottom of the page
+# Copyright and license strings; shown at the bottom of the page
```
Question: Is there a licence string in this block that I'm failing to see? As far as I can see, the license string is in the html_context far below.

medubelko replied:
This was from an earlier change. I'll revert it.

medubelko replied:
216fd98

izmalk reviewed (approved):
I like how it looks now!
However, I’m a bit concerned about the number of invisible changes — including some refactoring — that could potentially break things (even though everything seems fine at the moment). There also appear to be a few changes that might not be strictly necessary (see comments below).
I guess we can merge it now and do the code refactoring later, next cycle.

medubelko committed:
theme: revert comment about copyright string
